cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/k8s/UserServiceProgramMain.java

```diff
@@ -0,0 +1,379 @@
+
+    controllerFuture = new CompletableFuture<>();
+    programCompletion = new CompletableFuture<>();
+    maxStopSeconds = cConf.getLong(Constants.AppFabric.PROGRAM_MAX_STOP_SECONDS, 60);
+
+    programRunner = injector.getInstance(ServiceProgramRunner.class);
+
+    // fetch the app artifact
+    ArtifactFinder artifactFinder = injector.getInstance(ArtifactFinder.class);
+
+    NamespaceId programNamespace = programRunId.getNamespaceId();
+    Set<co.cask.cdap.api.artifact.ArtifactId> pluginArtifacts = new HashSet<>();
+    for (Plugin plugin : appSpec.getPlugins().values()) {
+      pluginArtifacts.add(plugin.getArtifactId());
+    }
+
+    // need to set the plugins directory in system args, as well as the host to bind to
+    File pluginsDir = new File(options.getTmpDir(), "plugins");
+    Map<String, String> systemArgs = new HashMap<>(programOptions.getArguments().asMap());
+    systemArgs.put(ProgramOptionConstants.PLUGIN_DIR, pluginsDir.getAbsolutePath());
+    systemArgs.put(ProgramOptionConstants.HOST, options.getBindAddress());
+    programOptions = new SimpleProgramOptions(programRunId.getParent(), new BasicArguments(systemArgs),
+                                              programOptions.getUserArguments());
+
+    try {
+      File programDir = new File(options.getTmpDir(), "program");
+      DirUtils.mkdirs(programDir);
+      File programJarFile = new File(programDir, "program.jar");
+      Location programJarLocation =
+        artifactFinder.getArtifactLocation(convert(programNamespace, appSpec.getArtifactId()));
+      Files.copy(programJarLocation.getInputStream(), programJarFile.toPath());
+
+      DirUtils.mkdirs(pluginsDir);
+      for (co.cask.cdap.api.artifact.ArtifactId artifactId : pluginArtifacts) {
+        Location location = artifactFinder.getArtifactLocation(convert(programNamespace, artifactId));
+        File destFile = new File(pluginsDir, String.format("%s-%s-%s.jar", artifactId.getScope().name(),
+                                                           artifactId.getName(), artifactId.getVersion().getVersion()));
+        Files.copy(location.getInputStream(), destFile.toPath());
+      }
+      programRunner = injector.getInstance(ServiceProgramRunner.class);
+
+      // Create the Program instance
+      programJarLocation = Locations.toLocation(programJarFile);
+      // Unpack the JAR file
+      BundleJarUtil.unJar(programJarLocation, programDir);
+      program = Programs.create(injector.getInstance(CConfiguration.class), programRunner,
+                                new ProgramDescriptor(programOptions.getProgramId(), appSpec), programJarLocation,
+                                programDir);
+    } catch (ArtifactNotFoundException e) {
+      throw new RuntimeException("Unable to find application artifact " + appSpec.getArtifactId(), e);
+    } catch (IOException e) {
+      Throwables.propagate(e);
+    }
+  }
+
+  @Override
+  public void start() {
```

Comment:
Instead of allowing `start` and `stop` to be overridable, I think it's better to encapsulate the start/stop logic into a single `AbstractIdleService`. It will make the code cleaner and also no need to try catch the exception and rethrow.